jwtmiddleware.go

```diff
@@ -26,95 +61,132 @@ type TokenExtractor func(r *http.Request) (string, error)
 
-// Options is a struct for specifying configuration options for the middleware.
-type Options struct {
-	// Validate handles validating a token.
-	Validate ValidateToken
+type JWTMiddleware struct {
+	// validateToken handles validating a token.
+	validateToken ValidateToken
 	// The name of the property in the request where the user information
 	// from the JWT will be stored.
 	// Default value: "user"
-	UserProperty string
+	contextKey string
 	// The function that will be called when there are errors in the
 	// middleware.
 	// Default value: OnError
-	ErrorHandler errorHandler
+	errorHandler ErrorHandler
 	// A boolean indicating if the credentials are required or not
 	// Default value: false
-	CredentialsOptional bool
+	credentialsOptional bool
 	// A function that extracts the token from the request
 	// Default: FromAuthHeader (i.e., from Authorization header as bearer token)
-	Extractor TokenExtractor
+	tokenExtractor TokenExtractor
 	// Debug flag turns on debugging output
 	// Default: false
-	Debug bool
-	// When set, all requests with the OPTIONS method will use authentication
-	// Default: false
-	EnableAuthOnOptions bool
+	debug bool
+	// When set, all requests with the OPTIONS method will have their
+	// token validated.
+	// Default: true
+	validateOnOptions bool
 }
 
-type JWTMiddleware struct {
-	Options Options
+// Option is how options for the middleware are setup.
+type Option func(*JWTMiddleware)
+
+func WithValidateToken(vt ValidateToken) Option {
+	return func(m *JWTMiddleware) {
+		m.validateToken = vt
+	}
 }
 
-func OnError(w http.ResponseWriter, r *http.Request, err string) {
-	http.Error(w, err, http.StatusUnauthorized)
+func WithContextKey(k string) Option {
+	return func(m *JWTMiddleware) {
+		m.contextKey = k
+	}
 }
 
-// New constructs a new Secure instance with supplied options.
-func New(options ...Options) *JWTMiddleware {
+func WithErrorHandler(h ErrorHandler) Option {
+	return func(m *JWTMiddleware) {
+		m.errorHandler = h
+	}
+}
 
-	var opts Options
-	if len(options) == 0 {
-		opts = Options{}
-	} else {
-		opts = options[0]
+func WithCredentialsOptional(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.credentialsOptional = value
 	}
+}
 
-	if opts.UserProperty == "" {
-		opts.UserProperty = "user"
+func WithTokenExtractor(e TokenExtractor) Option {
+	return func(m *JWTMiddleware) {
+		m.tokenExtractor = e
 	}
+}
 
-	if opts.ErrorHandler == nil {
-		opts.ErrorHandler = OnError
+func WithDebug(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.debug = value
 	}
+}
+
+func WithValidateOnOptions(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.validateOnOptions = value
+	}
+}
 
-	if opts.Extractor == nil {
-		opts.Extractor = FromAuthHeader
+// New constructs a new Secure instance with supplied options.
+func New(opts ...Option) *JWTMiddleware {
+	m := &JWTMiddleware{
+		validateToken:       func(string) (interface{}, error) { panic("not implemented") },
```

Comment:
Before v2 release we will either default to the josev2 for this, or require a specific validate token func to be passed into `New`.